In [49]:
import csv  
import re
import pandas as pd

ModuleNotFoundError: No module named 'pandas'

# Data Cleaning:
Pour télécharger les datas pour ce notebook, vous pouvez utilisez le [script associé](/nautilus_nlp/data/external/get_language_dataset.sh)

In [8]:
with open('../data/external/wili/x_test.txt',encoding='utf-8') as fp:
    test=fp.readlines()
    test=[doc.strip('\n') for doc in test]

Le Dataset Wili est un dataset d'identification de langue basé sur Wikipedia. Des exemples du tests set suivent: 

In [34]:

test[0]

"Ne l fin de l seclo XIX l Japon era inda çconhecido i sótico pa l mundo oucidental. Cula antroduçon de la stética japonesa, particularmente na Sposiçon Ounibersal de 1900, an Paris, l Oucidente adquiriu un apetite ansaciable pul Japon i Heiarn se tornou mundialmente coincido pula perfundidade, ouriginalidade i sinceridade de ls sous cuntos. An sous radadeiros anhos, alguns críticos, cumo George Orwell, acusórun Heiarn de trasferir sou nacionalismo i fazer l Japon parecer mais sótico, mas, cumo l'home qu'oufereciu al Oucidente alguns de sous purmeiros lampeijos de l Japon pré-andustrial i de l Período Meiji, sou trabalho inda ye balioso até hoije."

In [28]:
test[1]

'Schiedam is gelegen tussen Rotterdam en Vlaardingen, oorspronkelijk aan de Schie en later ook aan de Nieuwe Maas. Per 30 april 2017 had de gemeente 77.833 inwoners (bron: CBS). De stad is vooral bekend om haar jenever, de historische binnenstad met grachten, en de hoogste windmolens ter wereld.'

In [29]:
test[2]

'ГIурусаз батальонал, гьоркьор гIарадабиги лъун, ункъбокIон (каре) гьабун чIезарун руго. ТIаде гIарададул сачмаги гуллаги байдал, АхIмадханил бо тIурун буго. ГIумаханас цойгидал боязе тIаде кIанцIизе буюрухъ кьун буго. Гьезулги жо ккун гьечIо. Цинги живго ГIумахан кIанцIун вуго тушманасде тIаде («угъузилал рачун, дайтилал рачун, маххулъан бер баккун хунз цадахъ рачун»). Нахъе къалел ругел магIарулазда гьес гьарулеб букIун буго: «Нужеца яхI бахъе, гIолохъаби! Нилъеда данде гьал чIоларо», – ян.'

In [30]:
test[3]

'ರಾಜ್ಯಶಾಸ್ತ್ರದ ಪಿತಾಮಹೆ ಅರಿಸ್ಟಾಟಲ್. ರಾಜ್ಯಶಾಸ್ತ್ರದ ವೈಜ್ನಾನಿಕವಾದ್ ಅದ್ಯಾಯನ ಮಲ್ದಿನಾರ್ ಅರಿಸ್ಟಾಟಲ್. ಮೇರ್ 158 ರಾಷ್ಟ್ರದ ಸಂವಿಧಾನಲೆನ್ ಅರ್ಥ ಮಲ್ತೊನ್ದ್ the politics ಕೃತಿ ರಚನೆ ಮಲ್ದೆರ್. ಮೇರ್ ಕ್ರಿ.ಪೂ 387 ಗ್ರೀಕ್ ಸ್ಟಾಗಿರಡ್ ಜನಿಸಿಯೆರ್. ಮೆರೆನ ಅಮ್ಮೆರ್ ಮೆಸೆದೊನಿಯ ಅರಸೆರ್ನ ರಾಜವೈದ್ಯರ್ ಅದುದ್ ಇತ್ತೆರ್. ಎಲ್ಳಡೆ ಮೆರ್ ತತ್ವಶಾಸ್ತ್ರ,ಅರ್ಥಶಾಸ್ತ್ರ,ಸಂಗೀತ,ವಿಜ್ನಾನ,ಪೌರನೀತಿ,ಗಣಿತ ನೆಟ್ಟ್ ಪರಂಗತೆರ್ ಅದುದ್ ಇತ್ತೆರ್. ವಿದ್ಯಾಭ್ಯಾಸ ಮುಗಿ ಬೊಕ್ಕ ಪ್ಲೇಟೋ ಗ್ರೀಕ್ ತತ್ವಜ್ನಾನಿನೊಟ್ ಸೆರೊಂಡೆರ್. ಪ್ಲೇಟೋನ ಅಕಾಡೆಮಿಡ್ ಮಸ್ತ್ ವರ್ಷ ಬೇಲೆ ಮಲ್ತೆರ್. ಅಕಾಡೆಮಿಡ್ ನಿರ್ದೆಶೆಕೆರ್ನ ಹುದ್ದೆ ತಿಕ್ಕಂದೆ ಬೆಜರ್ ಅದುದ್ ಲೈಸಿಯಮ್ ಪನ್ಪಿನ ವಿಶ್ವವಿದ್ಯಾನಿಲಯ ಸ್ಥಾಪನೆ ಮಲ್ತೆರ್. ಬಿಕ್ಕ ಒಂತೆ ಸಮಯ ಅಲೆಗ್ಸಾಂಡರ್ ಚಕ್ರವರ್ತಿಗ್ ಗುರುವಾದ್ ಬೇಲೆ ಮಲ್ತೆರ್. ಮೇರ್ ಬರೆತಿನ ಪುಸ್ತಕ the politics,history of animals,metaphysics ಇತ್ಯಾದಿ.'

In [13]:
test[5]

'Тухайн үед дан ганц 20кг-ын Орос баллон хэрэглэдэг байсан уламжлалыг халж хэрэглэгчдийг сонголт ихтэй болгож 5,10,14,20 кг хэмжээтэй, баллонуудыг захиалан үйлдвэрлүүлж, өөрийн брэнд болгон хэрэглээнд нэвтрүүлж, хийн түлшээр ажилладаг тулга, төрөл бүрийн халаагуурууд, тоног төхөөрөмжийн ашиглалт, хийн түлшний ашиг тусыг хэвлэл мэдээллийн хэрэгслэлээр тасралтгүй сурталчилсаны үр дүнд 2002 оны 5 сар гэхэд хийн түлшний хэрэглээ сард 30 тн, 9 сард 50 тн хүрч өссөн юм.'

In [45]:
with open('../data/external/wili/y_test.txt',encoding='utf-8') as fp:
    y=fp.readlines()

In [47]:
with open('../data/external/wili/labels.csv',encoding='utf-8') as fp:
    labels=fp.readlines()

In [48]:
labels

['Label;English;Wiki Code;ISO 369-3;German;Language family;Writing system;Remarks;Synonyms\n',
 'ace;Achinese;ace;ace;Achinesisch;Austronesian;;;\n',
 'afr;Afrikaans;af;afr;Afrikaans;Indo-European;;;\n',
 'als;Alemannic German;als;gsw;Alemannisch;Indo-European;;(ursprünglich nur\xa0Elsässisch);\n',
 'amh;Amharic;am;amh;Amharisch;Afro-Asiatic;;;\n',
 'ang;Old English ;ang;ang;Altenglisch;Indo-European;;(ca. 450-1100);Angelsächsisch\n',
 'ara;Arabic;ar;ara;Arabisch;Afro-Asiatic;;;\n',
 'arg;Aragonese;an;arg;Aragonesisch;Indo-European;;;\n',
 'arz;Egyptian Arabic;arz;arz;Ägyptisch-Arabisch;Afro-Asiatic;;;\n',
 'asm;Assamese;as;asm;Assamesisch;Indo-European;;;\n',
 'ast;Asturian;ast;ast;Asturisch;Indo-European;;;\n',
 'ava;Avar;av;ava;Awarisch;Northeast Caucasian;;;\n',
 'aym;Aymara;ay;aym;Aymara;Aymaran;;;\n',
 'azb;South Azerbaijani;azb;azb;Südaserbaidschanisch;Turkic;Arabic;;\n',
 'aze;Azerbaijani;az;aze;Aserbaidschanisch;Turkic;Latin;;\n',
 'bak;Bashkir;ba;bak;Baschkirisch;Turkic;;;\n'

# Language identification

In [15]:
from nautilus_nlp.models import Fasttext_classifier 

Ici on va load le modèle pré-entrainer par Fasttext, disponible dans le folder data.
On peut utiliser soit le modèle quantizé (900ko) soit le classique

In [19]:
model = Fasttext_classifier.Fasttext_clf('../nautilus_nlp/data/lang_identification.ftz')

In [8]:
print_results(*model.test(valid_data))


N	100000
P@1	0.759
R@1	0.759


In [23]:
model.predict(test[3])


(('__label__kn',), array([0.99790311]))

Il faut donc 13 secondes pour entrainer un modèle sur 1.5 Millions de tweets, avec une précision de 0.759. Not bad

In [9]:
model.save_model('/home/nautilus_nlp/models/sentiments.bin')

In [11]:
!ls -lh '/home/nautilus_nlp/models/sentiments.bin'

-rw-r--r-- 1 root root 232M Mar 14 12:51 /home/nautilus_nlp/models/sentiments.bin


Ce modèle fait environ 230Mo

Maintenant si on quantize ce modèle

In [12]:
model.quantize(input=train_data, qnorm=True, retrain=True, cutoff=100000)
print_results(*model.test(valid_data))
model.save_model('/home/nautilus_nlp/models/sentiments.ftz')

N	100000
P@1	0.759
R@1	0.759


In [13]:
!ls -lh '/home/nautilus_nlp/models/sentiments.ftz'

-rw-r--r-- 1 root root 6.8M Mar 14 12:53 /home/nautilus_nlp/models/sentiments.ftz
